### ISSUE TITLE
Cannot download NSPL and Local Authority Districts data

### ISSUE THREAD

**Issue by cliffxuan** [ministryofjustice/postcodeinfo#155]
The http://geoportal.statistics.gov.uk website probably had some changes and now the urls for NSPL and Local Authority Districts do not work any more.

The data is still available on the website, but an automated way to get the url reliably isn't clear. 

These are the links. In fact, they are both on a different domain.

Local Authority Districts (February 2016): 
http://ons.maps.arcgis.com/home/item.html?id=c8cf971c001442f0bc665453fe1e7106

National Statistics Postcode Lookup ((February 2017):
http://ons.maps.arcgis.com/home/item.html?id=a26683d2393743f4b87c89141cd1b2e8

**Comment by cliffxuan** [comment]
Technical support of ONS pointed me to the new dev portal, but I'm not able to find the dataset we need.
https://web.ons.gov.uk/ons/apiservice/web/apiservice/home

**Comment by cliffxuan** [comment]
```
Good morning Cliff
 
The NSPL APIs for the latest version can be found here - http://geoportal.statistics.gov.uk/datasets/055c2d8135ca4297a85d624bb68aefdb_0
 
For a direct CSV download http://geoportal.statistics.gov.uk/datasets/055c2d8135ca4297a85d624bb68aefdb_0.csv
 
The local authorities direct CSV download can be retrieved here although this will change every year - http://geoportal.statistics.gov.uk/datasets/464be6191a434a91a5fa2f52c7433333_0.csv
 
I hope this helps.
 
We are always interested to hear about your use of our data.  If you have time please let us know.
 
Regards
 
Karen
Customer Services 
ONS Geography 
(01329) 44 4971
We are available daily 08.30 – 4pm
Contact us at ons.geography@ons.gsi.gov.uk 
Access Open Geography products at http://geoportal.statistics.gov.uk
Access Linked Data at http://statistics.data.gov.uk 
Follow us on Twitter @ONSgeography
```

**Comment by jrattle** [comment]
Here at Office of the Public Guardian, where we use the postcode look up tool for our online LPA service we have had lots of complaints from users. Would be great to get this fixed.

**Comment by cliffxuan** [comment]
another case:
```
---------- Forwarded message ---------
From: Rhys Williams <Rhys.Williams@hmcts.net>
Date: Thu, 30 Mar 2017 at 15:55
Subject: Re: HMCTS Postcode Info issue
To: Rob Stirling <rob.stirling@digital.justice.gov.uk>


Hi Rob,
 
Thanks for coming back to me. Apologies for the delay in responding, it’s been a hectic day.
 
Here are a few examples:
 
NG12 5GR is not returning 4 Rose Hill
ST12 9FB is not returning 4 Amherst Place
NG15 7EZ is not returning 87 Burberry Avenue
 
We have cross-checked these against the Royal Mail postcode finder.
 
Kind regards,
 
Rhys
```

**Comment by cliffxuan** [comment]
spent a good few hours with @aldavidson this afternoon and we had a temporary solution by manually downloading and importing the Address Base data.

a long term solution would need to do the following:

1. make sure if downloading and importing of one dataset fails, the other two should still work

2. use alternative data sources

NSPL (code update requires as it has different columns): https://data.gov.uk/dataset/national-statistics-postcode-lookup-uk

Local Authority: haven't found a data source yet. Even if we stick with current dataset, a fix is still needed as the column names are changed.

**Comment by aldavidson** [comment]
The import will take a few hours to complete,  we left it running at 5.30, I'll check it later tonight or first thing tomorrow

**Comment by aldavidson** [comment]
The import failed overnight again, only bringing in about 12m rows, where it should have around 26m. I've flipped the DNS to point to the prod1 stack, so it's back to the state is was in yesterday (~21m), while we investigate further on prod2

**Comment by aldavidson** [comment]
Looks like the AddressBase files that we get from Ordnance Survey have some invalid lines in:

```
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_001.csv
ERROR:  missing data for column "po_box_number"
CONTEXT:  COPY tmp_addressbase_import, line 887242: "151045080,"osgb1000002232240689",116462,"","""
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_002.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_003.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_004.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_005.csv
ERROR:  date/time field value out of range: "20"
HINT:  Perhaps you need a different "datestyle" setting.
CONTEXT:  COPY tmp_addressbase_import, line 769254, column la_start_date: "20"
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_006.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_007.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_008.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_009.csv
ERROR:  date/time field value out of range: "20"
HINT:  Perhaps you need a different "datestyle" setting.
CONTEXT:  COPY tmp_addressbase_import, line 852602, column rm_start_date: "20"
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_010.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_011.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_012.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_013.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_014.csv
ERROR:  unterminated CSV quoted field
CONTEXT:  COPY tmp_addressbase_import, line 690942: "10010232651,"osgb1000002148912480",51780845,"COSTA COFFEE","","","UNIT 1","WOOLWICH ARSENAL D L R ST..."
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_015.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_016.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_017.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_018.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_019.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_020.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_021.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_022.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_023.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_024.csv
ERROR:  missing data for column "rpc"
CONTEXT:  COPY tmp_addressbase_import, line 841288: "100110609728,"osgb1000002292439326",16001875,"","","","","",93,"","PORCHESTER DRIVE","CRAMLINGTON","..."
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_025.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_026.csv
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_027.csv
ERROR:  invalid input syntax for type date: "2"
CONTEXT:  COPY tmp_addressbase_import, line 800767, column la_start_date: "2"
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_028.csv
ERROR:  unquoted carriage return found in data
HINT:  Use quoted CSV field to represent carriage return.
CONTEXT:  COPY tmp_addressbase_import, line 53792
importing /tmp/addressbase_basic/AddressBase_FULL_2017-03-22_029.csv
```

Looking at the files manually, I can see that the last line of (...)001.csv seems to be cut off halfway through the line.

This makes the psql \COPY command (which we use for fast bulk import) drop the whole file if there are any errors in it - which accounts for the missing data.

I'll try manually downloading the files on my local laptop to see if the errors are there also, or if something has corrupted the files on the server

**Comment by aldavidson** [comment]
It gets better - the ftp server provided by O/S is currently empty, so i can't download the files locally (I have a vague memory of this happening before, when an update was nearly due - and the next one is due on April 27th)
 
When I try to download it via the website, you can't get the 29 zipped CSVs that we get from the FTP site, you have to download *10964* separate CSVs.

I have at least confirmed that re-extracting the CSVs from the downloaded zips reproduces the same error - so it's most likely to be an error in the original source rather than a corruption at our end. 

I will get onto O/S support this afternoon, and also try hacking the files in-situ to get rid of the corrupt lines - that should let us get back most of the missing data.

**Comment by cliffxuan** [comment]
re files not present on the ftp, I came across this issue around Christmas. WebOps told me the files are only available on the ftp for a short period of time when an update comes up.

@aldavidson in case you are not aware of, the zip files from the FTP are backed up on S3, so in case you need them, you can download from there.

**Comment by aldavidson** [comment]
file 28 has a large amount of non-printable characters at line 53792

**Comment by aldavidson** [comment]
@cliffxuan yeah, I looked, but the latest files on S3 in either prod bucket are from august 2016 unfortunately

**Comment by aldavidson** [comment]
OK, I've manually deleted the corrupted lines in the latest AddressBase files (see 
[address_base_FULL_2017-03-22_corrupt_lines.txt](https://github.com/ministryofjustice/postcodeinfo/files/947312/address_base_FULL_2017-03-22_corrupt_lines.txt) )
- and re-imported on prod2. It's successfully run through, and has 26.3m addresses.  This is still around 2.5m short of a full dataset, but is much better than where we were. Interestingly, all but one of the corrupted lines were the last line in that particular file, it looks like they've been truncated in the wrong place - so I suspect that accounts for the 2.5m difference.

So I'm going to switch the DNS back to prod2 and disable the crontab entry that runs download_and_import_all.sh, and we can re-group to think about a longer-term fix on Monday.

**Comment by aldavidson** [comment]
OK, done - 

NG12 5GR now returns 4 Rose Hill
ST12 9FB now returns 4 Amherst Place
NG15 7EZ now returns 87 Burberry Avenue

**Comment by cliffxuan** [comment]
awesome!

**Comment by aldavidson** [comment]
just checked the OS FTP site again - still no data

**Comment by aldavidson** [comment]
we finally have a new AddressBase release available on the OS FTP server, so I've set a manual import running on prod1 (the inactive stack). So far the files look to be much more consistent in size (28 x 199MB, plus a smaller 29th "tail end").

I'll report back on the results

**Comment by aldavidson** [comment]
I've checked the manual import run, and we have the full dataset:

```
postcodeinfo=> select count(*) from postcode_api_address;
  count   
----------
 28599838
(1 row)
```

So, I will:
# flip the DNS to make this stack live
# try a full download_and_import_all on the offline stack
# see what happens 

- if the addresses are imported successfully but we don't get the NSPL or local authorities, and it stays with the previous data for those two, I think that's still fine, and we should re-enable the job

**Comment by jrattle** [comment]
What was the outcome? Will it be stable in future?

**Comment by aldavidson** [comment]
@jrattle yes, it should be stable now. I've left the automatic download+import disabled on the live stack, and enabled on the offline replica. It ran on Friday, and the databases have identical numbers of all relevant records, so I think we're good.

Longer-term, we're now actively discussing moving the back-end of this over to using HMRC's service, which uses the premium AddressBase files, and does almost (but not quite) all that we need - ultimately it boils down to someone agreeing to pay for a developer to do the work. I'll keep you posted as and when decisions are made

**Comment by juder** [comment]
@aldavidson @cliffxuan 

Have you made any decisions about HMRC's service? 
We've received feedback that this postcode "BS25 1BU" doesn't work in the postcode look up tool. 
Do you know if Ordnance survey released another set of data and it fail? 

Sorry if this isn't the right place to leave this feedback or ask these questions, but not sure where the right place is now.

**Comment by aldavidson** [comment]
@juder we have ~28.5m addresses in the database at the moment, that's a full set, so there hasn't been a failed import.

I've checked the latest release of AddressBase files from Ordnance Survey (dated 28th April) and it doesn't contain any BS25 1BU postcodes, so the postcodeinfo service itself is behaving correctly in this case. Maybe that postcode has been discontinued recently, although I haven't managed to find any authoritative info on this. 

We are in the process of getting a dedicated team to work on this and port to HMRC's service if possible, I'll let you know when I have something more to report on that

**Comment by jrattle** [comment]
@aldavidson @cliffxuan Are there problems with the postcode look up again please? We use the address of MoJ as our test postcode and it keeps failing. SW1H 9AJ.

**Comment by aldavidson** [comment]
@jrattle curious - in the latest files released by Ordnance Survey (dated July 28th) the postcode 'SW1H 9AJ' does not appear, anywhere. The Ministry of Justice is listed as 52 Queen Annes Gate, SW1H 9AG. 

I don't know what the process is for querying OS's data with them, but I'll find out.

**Comment by jrattle** [comment]
@aldavidson thanks Al. That does seem odd. Please do let us know.

**Comment by jrattle** [comment]
@aldavidson we also checked a few other buildings including Portland Place, BBC broadcasting house, W1A 1AA and the address is missing. 
Not sure if this helps your query with OS.

**Comment by aldavidson** [comment]
Thanks, I've reported this to OS, and we'll see what happens.
